[compress]Default compress level use

ReyRen opened the pull request:
If the knet_compression_level not set or out of range, default
compression level specified seems better I think.
btw, thanks fabbione's suggestion: potentially use -1 as compress
module default, because barely any compress module currently accepts -1.
So, this also needs to trim a little bit in corosync.

Signed-off-by: yuan ren <yren@suse.com>

yuanren10 committed:
Default compress level use

Discuss with Fabio, invalid compression level not the knet
responsible for, so error logged. But if compress success but
dstLen larger than srcLen, defualt compression level will be
used, because the request level is not effective.

Signed-off-by: yuan ren <yren@suse.com>

fabbione commented:
@ReyRen thanks for the patch, but it doesn't really match what I had in mind when talking on IRC and the approach this patch is taking is "risky".

The current approach of passing "any" compress_level to the underlying libraries, without calling into val_level, is to avoid the risk of a library changing their ABI and knet being outdate the day after.

For example, library compress foo currently accepts a compress_level range 0 to 10. A new version of library compress foo accepts 0 to 11.

If we hardcode those checks in knet, for each compress library update, we need to update knet and that becomes a major headache for us and the different distro maintainers.

The current "almost free" approach delegate the issue of validating the compress_level to the user and the library, making knet almost transparent in that respect.

There is currently only one exception that is lzo* support because it requires a very specific mapping between level and calls into the library API (as you already know and have seen in the code).

What I had in mind when we were talking about using the default level was:

- accept user compress_level
- test it (compress_lib_test)
- if test fails due to invalid compress_level then use known default and re-test it, as it might have changed in the lower level libs
- if the second test works, then use default, otherwise error out.

Now, I understand your concern that some compress level produce no compression at all and they are not ideal, but it's the end user responsibility IMHO to pass valid data. knet can only make sure that the library underneath is ok with that value by testing it.

Make sense?

Fabio

PS coding style:
i would appreciate if you can change this:
if (compress_level < 0 || compress_level > 9) {
to this:
if ((compress_level < 0) || (compress_level > 9)) {

in lzo2 compress, you add the same line many times, you could set that as default and override only in the default: case (for example).

ReyRen commented:
@fabbione I really appreciate your patience in pointing out those problems. It really makes sense to me. I'd like to close this PR and will open a new one after modified.

fabbione commented:
@ReyRen or simply force push to this one.. as you prefer. For me it's the same :)

fabbione commented:
@ReyRen there is another improvement that we could add.

The compress_lib_test uses a 64K buffer filled with 0. I can't think of a single compress method that should not be able to reduce that to a few hundred bytes.

If the output of compress_lib_test is equal or bigger than the original buffer, then we know that the requested compress_level is useless. I don't think I want knet to change the value but we can at least generate a warning in the logs.

Make sense?

Fabio

fabbione commented:
@ReyRen we are getting closer :)
The only thing I don't like in this patch is this compress_def_level() api call that's nothing more than a wrapper for a compress.

I think I would prefer to have instead an API call to get_default_level from the module and don't duplicate the compress across, but just handle it all in compress.c.

Having get_defaut_level might be useful in future, for example we could have an external API for the users: knet_handle_get_default_compress_level(....) so that users can get and use the value.

ReyRen commented:
> Having get_defaut_level might be useful in future, for example we could have an external API for the users: knet_handle_get_default_compress_level(....) so that users can get and use the value.

Agree with you.

fabbione commented on libknet/compress.c:
```diff
@@ -209,6 +209,28 @@ static int compress_lib_test(knet_handle_t knet_h)
 		log_err(knet_h, KNET_SUB_COMPRESS, "Unable to compress test buffer. Please check your compression settings: %s", strerror(savederrno));
 		errno = savederrno;
 		return -1;
+	} else if ((long unsigned int)dst_comp_len >= KNET_DATABUFSIZE) {
```
hopefully this cast won't make CI unhappy... let see.

fabbione commented on libknet/compress.c:
```diff
@@ -209,6 +209,28 @@ static int compress_lib_test(knet_handle_t knet_h)
+		/*
+		 * compress not effect, try again using default compression level
+		 */
+		log_warn(knet_h, KNET_SUB_COMPRESS, "Useless compression level: %d", knet_h->compress_level);
```
I would prefer a more meaningful message here.

/*
 * compress not effective, try again using default compression level when available
 */
log_warn( .... "Requested compression level (%d) did not generate any compressed data (source: %zu destination: %zu)".. and add source and destination length to the message for debugging purposes.

fabbione commented on libknet/compress.c:
```diff
@@ -209,6 +209,28 @@ static int compress_lib_test(knet_handle_t knet_h)
+
+		if ((!compress_modules_cmds[knet_h->compress_model].ops->get_default_level()) ||
+				(knet_h->compress_level = compress_modules_cmds[knet_h->compress_model].ops->get_default_level() < 0)) {
+                        log_err(knet_h, KNET_SUB_COMPRESS, "Get default compression level failed");
```
Please expand the message a bit here.
"compression %s does not provide a default value", compress_module_name....

fabbione commented on libknet/compress.c:
```diff
@@ -209,6 +209,28 @@ static int compress_lib_test(knet_handle_t knet_h)
+			dst_comp_len = KNET_DATABUFSIZE_COMPRESS;
+			if (compress_modules_cmds[knet_h->compress_model].ops->compress(knet_h, src, KNET_DATABUFSIZE, dst, &dst_comp_len) < 0) {
+                                savederrno = errno;
+                                log_err(knet_h, KNET_SUB_COMPRESS, "Unable to compress withing default compression level: %s", strerror(savederrno));
```
"Unable to compress with default compression level....."

fabbione commented on libknet/compress.c:
```diff
@@ -209,6 +209,28 @@ static int compress_lib_test(knet_handle_t knet_h)
+				errno = savederrno;
+				return -1;
+			}
+			log_warn(knet_h, KNET_SUB_COMPRESS, "Default compression level: %d used", knet_h->compress_level);
```
"Requested compression level (%d) did not work, switching to default (%d)"

fabbione commented on libknet/compress_lz4.c:
```diff
@@ -15,6 +15,8 @@
 #include "logging.h"
 #include "compress_model.h"
 
+#define ACCELERATION_DEFAULT 1
```
I'll make this comment only once, but for each default, please add a comment from where the default is coming from.
For example:
ACCELERATION_DEFAULT 1 was found in man page X from library foo version 1.x
or
LZ4HC_DEFAULT was found in header file bla.h from library bar version 2.x

also, for all those default, we should have an #ifdef around them
#ifdef LZ4HC .....
#define KNET_COMPRESS_DEFAULT LZ4HC
#else 
#define KNET_COMPRESS_DEFAULT KNET_COMPRESS_UNKNOWN_DEFAULT
#endif

fabbione commented on libknet/compress_lzo2.c:
```diff
@@ -17,6 +17,7 @@
 #include "logging.h"
 #include "compress_model.h"
 
+
```
stray new line

fabbione commented on libknet/compress_model.h:
```diff
@@ -11,7 +11,14 @@
 #include "internals.h"
 
-#define KNET_COMPRESS_MODEL_ABI 1
+#define KNET_COMPRESS_MODEL_ABI            1
```
good spot here... we are changing the internal module API/ABI by adding the new call to get the default value. this should become 2.

fabbione commented on libknet/compress.c:
```diff
@@ -209,6 +210,32 @@ static int compress_lib_test(knet_handle_t knet_h)
+				(source: %zu destination: %zu)", request_level, sizeof(src), dst_comp_len);
+
+		if ((!compress_modules_cmds[knet_h->compress_model].ops->get_default_level()) ||
+				(knet_h->compress_level = compress_modules_cmds[knet_h->compress_model].ops->get_default_level() < 0)) {
```
Instead of using < 0 check if it's == KNET_COMPRESS_UNKNOWN_DEFAULT so we are free to change values around if we need.

fabbione commented:
test this please

fabbione commented:
@ReyRen so beside those super minor nit-picks, we should probably introduce some tests to verify that this code works as expected. I am thinking that we could extend api_knet_handle_compress test to use some "bad" compress_value and verify that we get back the default. Make sense?

good work btw, I am very happy how the code is evolving.

fabbione commented on libknet/compress_model.h:
```diff
@@ -11,7 +11,14 @@
+#define KNET_COMPRESS_MODEL_ABI            1
+#define KNET_COMPRESS_UNKNOWN_DEFAULT    (-1)
+#define LZ4_COMPRESS_DEFAULT               1 /* lz4 default compression level from lz4.c */
+#define ZLIB_COMPRESS_DEFAULT              6 /* zlib default compression level from zlib.h */
```
one more question here.. for example looking at zlib.h, I can see default compression is:
Z_DEFAULT_COMPRESSION (-1) that as the header says is 6.
I think it would make sense to use the header #define for this.

For example:
#define ZLIB_COMPRESS_DEFAULT Z_DEFAULT_COMPRESSION

this would require moving those defines back into the modules, but the benefit is that if a library changes default, then we can pick it up at the next build without having to change the hardcoded numbers here.

Make sense?

ReyRen commented:
> @ReyRen so beside those super minor nit-picks, we should probably introduce some tests to verify that this code works as expected. I am thinking that we could extend api_knet_handle_compress test to use some "bad" compress_value and verify that we get back the default. Make sense?

Yes, make sense

fabbione commented:
add to whitelist

fabbione committed:
[tests] ignore libnss errors from OpenSuse Tumbleweed

Signed-off-by: Fabio M. Di Nitto <fdinitto@redhat.com>

yuanren10 committed:
[compress]Default compression level use

1. add test casees for a module without default.Using default
compression level.
2. Discuss with Fabio, invalid compression level not the knet
responsible for, so error logged. But if compress success but
dstLen larger than srcLen, defualt compression level will be
used, because the request level is not effective.

Signed-off-by: yuan ren <yren@suse.com>
